README.md
# Forrst

**IMPORTANT:** this library is still in the early stages of development. More details will
be added once it reaches a stable version.

forrst is a Ruby library for the [Forrst][forrst] API. This implementation aims to become
an implementation that focuses on ease of use, minimal amount of dependencies and
stability. I'm also aiming for support of all Ruby distributions such as JRuby and MRI.

## Example

    require 'rubygems'
    require 'forrst'

    Forrst.configure do |config|
      config.id           = '123'
      config.secret       = '123'
      config.access_token = '123'
    end

    # Get a single user
    user = Forrst::User['YorickPeterse']
    user.name    # => 'Yorick Peterse'
    user.twitter # => 'YorickPeterse'

    # Get a bunch of posts
    posts = Forrst::Post.find(:type => :code, :sort => :popular)
    posts.each do |post|
      puts post.title
      puts post.user.username
    end

    # Or get a single post by it's ID or tiny ID
    post = Forrst::Post[123]
    post = Forrst::Post['YJp']

    post.title # => 'Some post title'

    post.comments.each do |comment|
      puts comment.body
    end

## Requirements & Installation

* Ruby 1.8.7, Rubinius, JRuby or Ruby 1.9.2 (1.9.1 and 1.9 are untested)
* An authenticated application (not available at the moment)
* RVM (when developing the library itself)

Installing the gem can be done in a single command:

    $ gem install forrst

If you want to submit pull requests or just check out the source code you should run the
following commands instead:

    $ git clone git://github.com/YorickPeterse/Forrst.git forrst
    $ cd forrst
    $ rvm gemset import .gems
    $ rake test

The command `rake test` is not required but recommended as it will check to see if your
environment is running the tests properly.

## Contributing

Contributions are more than welcome as long as they follow the following guidelines:

* Don't break compatibility with Ruby distributions such as JRuby and Rubinius.
* Document your code.
* Use Unix line endings (`\n`)
* Don't add bullshit code just because you like the way it's named or because you're used
  to it. New code (or modified code) should be useful.

## License

This API is licensed under the MIT license. A copy of this license can be found in the
file "LICENSE".

[forrst]: http://forrst.com/


CHANGELOG.md
# Changelog

## 0.1.1 - June 5th, 2011

The first public release of the library.


spec/helper.rb
require File.expand_path('../../lib/forrst', __FILE__)
require 'rspec'

Faraday.default_adapter = :test

def get_fixture(name)
  path = File.expand_path("../fixtures/#{name}.json", __FILE__)
  return File.read(path, File.size(path))
end

RSpec.configure do |c|
  c.fail_fast = true
end

# Configure the library so we don't have to do this over and over again
Forrst.configure do |config|

end

# Set up the stub requests so we don't bombard the Forrst server
Forrst.oauth.connection = Faraday.new(Forrst.oauth.site)

Forrst.oauth.connection.build do |builder|
  builder.adapter(:test) do |stub|
    stub.get(Forrst::StatisticsURL) do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('stats')]
    end

    stub.get(Forrst::User::InfoURL + '?id=6998') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('user/info')]
    end

    stub.get(Forrst::User::InfoURL + '?username=YorickPeterse') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('user/info')]
    end

    stub.get(Forrst::User::PostsURL + '?username=YorickPeterse') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('user/posts')]
    end

    stub.get(Forrst::Post::ListURL + '?post_type=code') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/list_code')]
    end

    stub.get(Forrst::Post::ListURL + '?post_type=code&sort=popular') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/list_popular')]
    end

    stub.get(Forrst::Post::ShowURL + '?id=86427') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/show')]
    end

    stub.get(Forrst::Post::ShowURL + '?tiny_id=YPj') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/show')]
    end

    stub.get(Forrst::Post::AllURL) do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/all')]
    end

    stub.get(Forrst::Post::AllURL + '?after=86541') do |env|
      [200, {'Content-Type' => 'application/json'}, get_fixture('posts/all_after')]
    end
  end
end


spec/forrst/all.rb
require File.expand_path('../../helper', __FILE__)
require 'rspec/autorun'

['monkeys/hash', 'forrst', 'user', 'post'].each do |spec|
  require File.expand_path("../#{spec}", __FILE__)
end


spec/forrst/post.rb
require File.expand_path('../../helper', __FILE__)

describe('Forrst::Post') do

  it('Get a list of all code posts') do
    posts = Forrst::Post.find(:type => :code)

    posts.size.should                === 10
    posts[0].type.should             === 'code'
    posts[0].post_id.should          === 86456
    posts[0].created_at.class.should == Date
    posts[0].user.username.should    === 'bguillermo'
  end

  it('Get a list of all code posts sorted by popularity') do
    posts = Forrst::Post.find(:type => :code, :sort => :popular)

    posts.size.should                === 4
    posts[0].type.should             === 'code'
    posts[0].post_id.should          === 86034
    posts[0].user.username.should    === 'amatyr4n'
    posts[0].updated_at.class.should == Date
  end

  it('Retrieve a single post by it\'s ID') do
    post = Forrst::Post[86427]

    post.user.username.should === 'YorickPeterse'
    post.title.should         === 'Forrst API in Ruby'
    post.type.should          === 'link'
  end

  it('Retrieve a single post by it\'s tiny ID') do
    post = Forrst::Post['YPj']

    post.user.username.should === 'YorickPeterse'
    post.title.should         === 'Forrst API in Ruby'
    post.type.should          === 'link'
    post.link?.should         === true
  end

  it('Retrieve all posts') do
    posts = Forrst::Post.all

    posts.size.should             === 7
    posts[0].post_id.should       === 86541
    posts[0].user.username.should === 'jtkendall'
  end

  it('Retrieve all posts after a certain ID') do
    posts = Forrst::Post.all(:after => 86541)

    posts.size.should             === 7
    posts[0].post_id.should       === 86540
    posts[0].user.username.should === 'ericandrewscott'
  end

end


lib/forrst.rb
require 'rubygems'
require 'oauth2'
require 'json'
require 'date'

['version', 'monkeys/hash', 'user', 'post', 'comment'].each do |file|
  require File.expand_path("../forrst/#{file}", __FILE__)
end

##
# The Forrst gem is an API library for the Forrst API that focuses on stability, ease of
# use and minimal dependencies.
#
# @author Yorick Peterse
# @since  0.1a
#
module Forrst
  ##
  # The URL to the API endpoint.
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  URL = 'http://forrst.com/api/v2/'

  ##
  # URL relative to Forrst::URL that contains all the API statistics.
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  StatisticsURL = '/stats'

  ##
  # A string containing the date format used for all dates returned by the API.
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  DateFormat = '%Y-%m-%d %H:%M:%S'


  class << self
    ##
    # The access token returned once a client has been authorized.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_accessor :access_token

    ##
    # The ID of the application as provided by Forrst.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_accessor :id

    ##
    # The secret of the application as provided by Forrst.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_accessor :secret

    ##
    # Instance of OAuth2::Client.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_accessor :oauth

    ##
    # The user agent to use for all API calls.
    #
    # @author Yorick Peterse
    # @since  0.1.2
    #
    attr_accessor :user_agent

    ##
    # Sets various configuration options in the module so that they can be used by other
    # parts of this gem.
    #
    # @example
    #  Forrst.configure do |client|
    #    client.access_token = '1234abc'
    #  end
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @yield  self
    #
    def configure
      yield self

      @oauth = OAuth2::Client.new(@id, @secret, :site => URL)
    end

    ##
    # Sends a request to the API server using the method, URL and optionally a hash of
    # options.
    #
    # @author Yorick Peterse
    # @since  0.1.2
    # @param  [Symbol] verb The HTTP verb to use as a symbol.
    # @param  [String] url The URI relative to the main one to send the request to.
    # @param  [Hash] options A hash with additional options to pass to Faraday.
    # @see    OAuth2::Client#request
    # @return [Mixed]
    #
    def request(verb, url, options = {})
      @user_agent ||= "Forrst/#{Forrst::Version} (#{RUBY_DESCRIPTION})"

      # Be a nice guy and set those headers correctly
      Forrst.oauth.connection.headers = {
        'User-Agent' => @user_agent, 
        'Accept'     => 'application/json, text/javascript'
      }

      return Forrst.oauth.request(verb, url, options)
    end

    ##
    # Gets a set of statistics from the API server.
    #
    # @example
    #  stats = Forrst.statistics
    #  stats[:limit] # => 150
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [Hash]
    #
    def statistics
      response = Forrst.request(:get, StatisticsURL)
      response = JSON.load(response)
      hash     = {
        :limit => response['resp']['rate_limit'].to_i,
        :calls => response['resp']['calls_made'].to_i
      }
    
      return hash
    end
  end # class << self
end # Forrst


lib/forrst/user.rb
module Forrst
  ##
  # Forrst::User is the class used for retrieving details about a particular user. A user
  # is retrieved by calling Forrst::User[] and passing a username or ID to it:
  #
  #     user = Forrst::User['yorickpeterse'] => #<Forrst::User:0x1234567>
  #
  # Once an instance has been created you can retrieve various details, such as the
  # username, simply by calling methods on the resulting object:
  #
  #     user.username => 'YorickPeterse'
  #
  # When retrieving a user you can retrieve the user's posts by calling Forrst::User#posts
  # as following:
  #
  #     user.posts.each do |post|
  #       puts post.title
  #     end
  #
  # When retrieving the user's posts the library will use the class Forrst::Post which
  # allows you to retrieve post related data such as all the comments:
  #
  #     user.posts.each do |post|
  #       post.comments.each do |comment|
  #         puts comment.body
  #       end
  #     end
  #
  # Note that comments are lazy loaded. This means that for each iteration in the posts
  # array a GET request is fired to the Forrst API. Currently there's no way to work
  # around this so you'll have to be careful with using this method.
  #
  # This class also provides the following shortcut methods for figuring out the type of
  # user:
  #
  # * developer?
  # * designer?
  # * developer_and_designer?
  #
  # These can be used as following:
  #
  #     if user.developer?
  #       puts "It's a nerd!"
  #     elsif user.designer?
  #       puts "It's a hippie!"
  #     end
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  class User
    ##
    # URL relative to Forrst::URL for retrieving user information.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    InfoURL = '/users/info'

    ##
    # URL relative to Forrst::URL for retrieving the posts of a user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    PostsURL = '/users/posts'

    ##
    # The user's ID.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :user_id

    ##
    # The username as used on Forrst.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :username

    ##
    # The real name of the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :name

    ##
    # The URL to the Forrst profile of the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :url

    ##
    # A hash containing various statistics such as the amount of comments, likes, etc.
    # Note that the keys of this hash are symbols, not strings.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :statistics

    ##
    # The description of the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :bio

    ##
    # The Twitter username of the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :twitter

    ##
    # A hash containing all the photos of the user. All the keys of this hash are symbols
    # just like the statistics hash.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :photos

    ##
    # An array containing the types of the user. Can be any of the following:
    #
    # * ['developer']
    # * ['designer']
    # * ['developer', 'designer']
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :type

    ##
    # The URL to the user's website.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :homepage

    ##
    # A boolean that indicates if the user is listed in the Forrst.me directory or not.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :listed

    ##
    # An array of tags for the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :tags

    ##
    # Retrieves a single user by it's username or ID and returns a new instance of
    # Forrst::User with these details.
    #
    # @example
    #  Forrst::User['yorickpeterse']
    #  Forrst::User[6998]
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [String/Fixnum] options
    # @return [Forrst::User]
    #
    def self.[](options)
      if options.class == String
        options = {:username => options}
      elsif options.class == Fixnum
        options = {:id => options.to_s}
      else
        raise(TypeError, "Expected Hash or Fixnum but got #{options.class} instead")
      end

      response = Forrst.request(:get, InfoURL, options)

      return User.new(response)
    end

    ##
    # Given a JSON response (as a string) this method will parse it using the JSON gem and
    # return a new instance of Forrst::User with all the details set.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [String/Hash] response A string containing a JSON response sent back by the
    # Forrst server or an instance of Hash in case the response has already been parsed.
    # @return [Forrst::User]
    #
    def initialize(response)
      if response.class != Hash
        response = JSON.load(response)
        response = response['resp']
      end

      # We're not directly setting Forrst::User#id as that will trigger warnings in both
      # JRuby and Rubinius.
      @user_id   = response['id']
      @username  = response['username']
      @name      = response['name']
      @url       = response['url']
      @homepage  = response['homepage_url']
      @listed    = response['in_directory']
      @twitter   = response['twitter']
      @bio       = response['bio']

      @statistics = {
        :comments  => response['comments'].to_i,
        :likes     => response['likes'].to_i,
        :followers => response['followers'].to_i,
        :following => response['following'].to_i,
        :posts     => response['posts'].to_i
      }

      # Photos come as a hash with rather wacky keys so those need to be changed as well.
      @photos = {
        :extra_large => response['photos']['xl_url'],
        :large       => response['photos']['large_url'],
        :medium      => response['photos']['medium_url'],
        :small       => response['photos']['small_url'],
        :thumbnail   => response['photos']['thumb_url']
      }

      # Tags aren't always present
      if response.key?('tag_string')
        @tags = response['tag_string'].split(',')
      else
        @tags = []
      end

      # Last but not least, the user type!
      @type = response['is_a'].split('&').map { |i| i.strip }
    end

    ##
    # Retrieves all the posts for the current user. Each post is an instance of
    # Forrst::Post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [Hash] options A hash with additional options to use when retrieving all the
    # posts.
    # @option options [String/Symbol] :type The type of posts to retrieve such as :code or
    # :question.
    # @option options [Fixnum] :limit The amount of posts to retrieve.
    # @option options [After]  :after An ID offset used for retrieving a set of posts
    # after the given ID.
    # @return [Array]
    #
    def posts(options = {})
      options  = {
        :username => @username
      }.merge(options.subset(:limit, :type, :after))

      response = Forrst.request(:get, PostsURL, options)
      response = JSON.load(response)

      return response['resp'].map do |post|
        Forrst::Post.new(post)
      end
    end

    ##
    # Checks if the user is a developer or not.
    #
    # @author Yorick Peterse
    # @since  0.1
    # @return [TrueClass/FalseClass]
    #
    def developer?
      return @type === ['developer']
    end

    ##
    # Checks if the user is a designer or not.
    #
    # @author Yorick Peterse
    # @since  0.1
    # @return [TrueClass/FalseClass]
    #
    def designer?
      return @type === ['designer']
    end

    ##
    # Checks if the user is a developer and a designer.
    #
    # @author Yorick Peterse
    # @since  0.1
    # @return [TrueClass/FalseClass]
    #
    def developer_and_designer?
      # The shorthand "return A and B" generates a void error so we'll have to do it this
      # way.
      if @type.include?('designer') and @type.include?('developer')
        return true
      else
        return false
      end
    end
  end # User
end # Forrst


lib/forrst/comment.rb
module Forrst
  ##
  # Forrst::Comment is a class used for individual comments.
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  class Comment
    ##
    # The ID of the comment.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :comment_id

    ##
    # An instance of Forrst::User containing the details of the user.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :user

    ##
    # The body of the comment.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :body

    ##
    # The date and time on which the comment was created.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :created_at

    ##
    # The date and time on which the comment was updated.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :updated_at

    ##
    # Creates a new instance of Forrst::Comment and parses the response (either in JSON or
    # as a has).
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [String/Hash] response The API response sent back from the server.
    #
    def initialize(response)
      if response.class != Hash
        response = JSON.load(response)
      end

      @comment_id = response['id']
      @user       = Forrst::User.new(response['user'])
      @body       = response['body']
      @created_at = Date.strptime(response['created_at'], Forrst::DateFormat)
      @updated_at = Date.strptime(response['updated_at'], Forrst::DateFormat)
    end
  end # Comment
end # Forrst


lib/forrst/post.rb
module Forrst
  ##
  # Forrst::Post is the class that represents a single post. Posts can be one of the
  # following types:
  #
  # * code
  # * snap
  # * question
  # * link
  #
  # Based on these types certain fields may not be set. For example, the description()
  # method will return a nil value in case the post type is set to "question" as those
  # don't get an extra description.
  #
  # Retrieving a single post or multiple posts can be done using Forrst::Post.[] and
  # Forrst::Post.find(). If you want to retrieve multiple posts you'll have to use find(),
  # if you only want to retrieve a single post you use []():
  #
  #     # Retrieves all code posts.
  #     Forrst::Post.find(:type => :code).each do |post|
  #       puts post.title
  #     end
  #
  #     # Retrieve's the post with ID 123
  #     post = Forrst::Post[123]
  #     puts post.title
  #
  # Each post will also contain the method comments(). This method sends an authenticated
  # GET request to the Forrst server to retrieve all comments for the specified posts.
  # This works as following:
  #
  #     post.comments.each do |comment|
  #       puts comment.body
  #     end
  #
  # Note that similar to Forrst::User#posts all comments are lazy-loaded as the API
  # provides no means of eager loading all comments. Use with care or you might risk
  # getting your ass blocked.
  #
  # Just like the class Forrst::User this class has a few shortcut methods for checking
  # the type of post:
  #
  # * code?
  # * snap?
  # * question?
  # * link?
  #
  # All of these will return true or false depending on the post type.
  #
  # @author Yorick Peterse
  # @since  0.1a
  #
  class Post
    ##
    # URL relative to Forrst::URL that contains a list of all posts.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    AllURL = '/posts/all'

    ##
    # URL relative to Forrst::URL that contains the details of a single post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    ShowURL = '/posts/show'

    ##
    # URL relative to Forrst::URL that contains a list of all posts.
    #
    # @author Yorick Peterse
    # @since  0.1
    #
    ListURL = '/posts/list'

    ##
    # URL relative to Forrst::URL that contains all the comments for a given post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    CommentsURL = '/posts/comments'

    ##
    # The ID of the post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :post_id

    ##
    # The "slug" that's added to the URL.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :tiny_id

    ##
    # The type of post ("code" for example).
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :type

    ##
    # The URL to the post on Forrst.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :post_url

    ##
    # The date and time on which the post was created. The value of this attribute is an
    # instance of the Date class.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :created_at

    ##
    # The date and time on which the post was updated. The value of this attribute is an
    # instance of the Date class.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :updated_at

    ##
    # The author of the post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :user

    ##
    # Boolean that indicates if the post has been published or not.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :published

    ##
    # Boolean that indicates if the post is a public post or requires users to log in.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :public

    ##
    # The title of the post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :title

    ##
    # If the post type was a link this field will contain the URL to the external page.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :url

    ##
    # The content of the post. If it's a question this field contains the question, if
    # it's a code post it will contain the code attached to the post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :content

    ##
    # Field containing the description of the post, not used for questions.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :description

    ##
    # The description in HTML.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :formatted_description

    ##
    # The content in HTML.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :formatted_content

    ##
    # A hash containing the number of comments, likes, etc.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :statistics

    ##
    # An array of all the tags used in the post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :tags

    ##
    # A hash containing the URLs to various sizes of the snap in case the post type is
    # "snap".
    #
    # @author Yorick Peterse
    # @since  0.1a
    #
    attr_reader :snaps

    ##
    # Given a custom set of options this method will retrieve a list of posts. It's
    # important to remember that the keys of the hash passed to this method should be
    # symbols and *not* strings.
    #
    # Note that this method will always return an array of posts, if you want to retrieve
    # a single post use Forrst::Post[] instead.
    #
    # @example
    #  Forrst::Post.find(:type => :question)
    #  Forrst::Post.find(:type => :code, :sort => :recent)
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [Hash] options
    # @option options [Symbol] :type The type of posts to retrieve. Can be one of
    # the following: code, snap, link or question.
    # @option options [Symbol] :sort The sort order to use, can be any of the
    # following: recent, popular or best.
    # @option options [Fixnum] :page The page to use. Each page contains a (currently
    # unknown) number of posts.
    # @return [Array]
    #
    def self.find(options)
      # Set the correct keys for the API call
      query_items             = {}
      query_items[:post_type] = options[:type].to_s if options.key?(:type)
      query_items[:sort]      = options[:sort].to_s if options.key?(:sort)
      query_items[:page]      = options[:page].to_s if options.key?(:page)

      response = Forrst.request(:get, ListURL, query_items)
      response = JSON.load(response)

      return response['resp']['posts'].map do |post|
        Post.new(post)
      end
    end

    ##
    # Retrieves a single post by it's ID or tiny ID. If the parameter given is a Fixnum
    # this method assumes is the ID, if it's a string it assumes it's the tiny ID.
    #
    # @example
    #  Forrst::Post[10].title
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [Fixnum/String] id The ID or the tiny ID.
    # @return [Forrst::Post]
    #
    def self.[](id)
      if id.class == Fixnum
        query_items = {:id => id}
      elsif id.class == String
        query_items = {:tiny_id => id}
      else
        raise(TypeError, "Got #{id.class} but expected Fixnum or String")
      end

      response = Forrst.request(:get, ShowURL, query_items)
      response = JSON.load(response)

      return Post.new(response['resp'])
    end

    ##
    # Gets all posts from the API, optionally after a certain ID.
    #
    # @example
    #  Forrst::Post.all.each do |post|
    #    puts post.title
    #  end
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [Hash] options A hash containing additional options to use for the request.
    # @option options [Fixnum] :after The ID used for the offset of all posts.
    # @return [Array]
    #
    def self.all(options = {})
      options  = options.subset(:after)
      response = Forrst.request(:get, AllURL, options)
      response = JSON.load(response)

      return response['resp']['posts'].map do |post|
        Post.new(post)
      end
    end

    ##
    # Creates a new instance of Forrst::Post and optionally decodes a JSON response. If a
    # response is already decoded it's data is merely assigned to the current instance.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @param  [String/Hash] response The API response in either JSON format or a hash.
    # @return [Forrst::Post]
    #
    def initialize(response)
      if response.class != Hash
        response = JSON.load(response)
      end

      @post_id     = response['id'].to_i
      @tiny_id     = response['id']
      @type        = response['post_type']
      @post_url    = response['post_url']
      @created_at  = Date.strptime(response['created_at'], Forrst::DateFormat)
      @updated_at  = Date.strptime(response['updated_at'], Forrst::DateFormat)
      @user        = Forrst::User.new(response['user'])
      @published   = response['published']
      @public      = response['public']
      @title       = response['title']
      @url         = response['url']
      @content     = response['content']

      @description           = response['description']
      @formatted_description = response['formatted_description']
      @formatted_content     = response['formatted_content']
      @tags                  = response['tags']

      @statistics = {
        :comments => response['comment_count'].to_i,
        :likes    => response['like_count'].to_i
      }

      # Get all the snaps
      @snaps = {}

      if response.key?('snaps')
        @snaps[:extra_large] = response['snaps']['mega_url']
        @snaps[:keith]       = response['snaps']['keith_url'] # The fuck is a keith?
        @snaps[:large]       = response['snaps']['large_url']
        @snaps[:medium]      = response['snaps']['medium_url']
        @snaps[:small]       = response['snaps']['small_url']
        @snaps[:thumbnail]   = response['snaps']['thumb_url']
        @snaps[:original]    = response['snaps']['original_url']
      end
    end

    ##
    # Retrieves all the comments for the current post.
    #
    # @example
    #  post = Forrst::Post[10]
    #  post.comments.each do |comment|
    #    puts comment.body
    #  end
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [Array]
    #
    def comments
      response = Forrst.request(:get, CommentsURL, :id => @post_id)
      response = JSON.load(response)

      return response['comments'].map do |comment|
        Forrst::Comment.new(comment)
      end
    end

    ##
    # Checks if the post is a question.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [TrueClass/FalseClass]
    #
    def question?
      return @type === 'question'
    end

    ##
    # Checks if the post is a snap.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [TrueClass/FalseClass]
    #
    def snap?
      return @type === 'snap'
    end

    ##
    # Checks if the post is a code post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [TrueClass/FalseClass]
    #
    def code?
      return @type === 'code'
    end

    ##
    # Checks if the post is a link post.
    #
    # @author Yorick Peterse
    # @since  0.1a
    # @return [TrueClass/FalseClass]
    #
    def link?
      return @type === 'link'
    end
  end # Post
end # Forrst)
